### PR TITLE
Add HistoricalCursor.getReadableOffset() to access unwrapped offset in selectors

### LINKED ISSUE #4628: Stack overflow in StringRuntimeShapeInspector
On master and 0.10.1-rc2, I'm seeing a stack overflow when I run the following query on the wikiticker data:

```
{
  "queryType": "topN",
  "dataSource": "wikiticker",
  "intervals": "1000/3000",
  "granularity": "all",
  "context": {
    "timeout": 10000,
    "useCache": false,
    "populateCache": false
  },
  "filter": {
    "type": "bound",
    "dimension": "commentLength",
    "ordering": "numeric",
    "lower": 46.64980229268867,
    "lowerStrict": true
  },
  "dimension": {
    "type": "default",
    "dimension": "channel",
    "outputName": "Channel"
  },
  "aggregations": [
    {
      "name": "Count",
      "type": "longSum",
      "fieldName": "added"
    }
  ],
  "metric": "Count",
  "threshold": 5
}
```

```
java.lang.StackOverflowError
	at java.lang.StringBuilder.append(StringBuilder.java:136) ~[?:1.8.0_51]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:93) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:145) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.segment.column.IndexedLongsGenericColumn.inspectRuntimeShape(IndexedLongsGenericColumn.java:107) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.appendHotLoopCalleeShape(StringRuntimeShape.java:107) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:95) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:145) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:136) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.segment.QueryableIndexStorageAdapter$CursorSequenceBuilder$1$1QueryableIndexBaseCursor$3.inspectRuntimeShape(QueryableIndexStorageAdapter.java:656) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.appendHotLoopCalleeShape(StringRuntimeShape.java:107) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:95) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:145) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:136) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.segment.filter.Filters$4.inspectRuntimeShape(Filters.java:464) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.appendHotLoopCalleeShape(StringRuntimeShape.java:107) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:95) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:145) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:136) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.segment.FilteredOffset.inspectRuntimeShape(FilteredOffset.java:130) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.appendHotLoopCalleeShape(StringRuntimeShape.java:107) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:95) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:145) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:136) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.segment.QueryableIndexStorageAdapter$CursorSequenceBuilder$1$1QueryableIndexBaseCursor$3.inspectRuntimeShape(QueryableIndexStorageAdapter.java:657) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.appendHotLoopCalleeShape(StringRuntimeShape.java:107) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:95) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:145) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:136) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.segment.filter.Filters$4.inspectRuntimeShape(Filters.java:464) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.appendHotLoopCalleeShape(StringRuntimeShape.java:107) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:95) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:145) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.query.monomorphicprocessing.StringRuntimeShape$Inspector.visit(StringRuntimeShape.java:136) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
	at io.druid.segment.FilteredOffset.inspectRuntimeShape(FilteredOffset.java:130) ~[druid-processing-0.11.0-SNAPSHOT.jar:0.11.0-SNAPSHOT]
```

### PR DESCRIPTION
**Pull request by leventov**
Fixes #4628 

@jon-wei could you please contribute a failing test?

### CODE REVIEW

**Comment by jon-wei** [comment]
Sure, I can take a look at that in a bit

**Comment by jon-wei** [comment]
@leventov this unit test will trigger the issue:

https://github.com/jon-wei/druid/commit/be53f9ed1d075439c3a98d7e9011816779d0af8e

File: processing/src/main/java/io/druid/segment/column/SimpleDictionaryEncodedColumn.java [hunk 1]
```diff
@@ -274,7 +274,7 @@ public void inspectRuntimeShape(RuntimeShapeInspector inspector)
         {
           inspector.visit("column", column);
           inspector.visit("offsetHolder", offsetHolder);
-          inspector.visit("offset", offsetHolder.getOffset());
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/column/SimpleDictionaryEncodedColumn.java]: I guess this line (and others like it) are the ones that actually fix the bug.

**Comment by leventov** [comment]
I think cycle detection won't win us anything, because it could fail only in runtime, but it fails with StackOverflowError already.

**Comment by gianm** [comment]
At least the error message would be clearer if it identified the cycle.

**Comment by leventov** [comment]
But supporting IdentityHashMap of already visited references has a non-trivial cost.

**Pull request merged**